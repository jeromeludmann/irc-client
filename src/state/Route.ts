import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export interface Routable extends Action<string> {
  route: Route;
}
