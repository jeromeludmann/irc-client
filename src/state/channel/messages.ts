import { SEND_INPUT_VALUE, SendInputValueAction } from "@app/actions/ui/input";
import { JoinAction, MESSAGE_JOIN } from "@app/actions/messages/join";
import { NoticeAction, MESSAGE_NOTICE } from "@app/actions/messages/notice";
import { PingAction, MESSAGE_PING } from "@app/actions/messages/ping";
import { PrivmsgAction, MESSAGE_PRIVMSG } from "@app/actions/messages/privmsg";

export type MessageListState = string[];

export type MessageListAction =
  | SendInputValueAction
  | JoinAction
  | NoticeAction
  | PingAction
  | PrivmsgAction;

export const messagesInitialState: MessageListState = [];

export default function reduceMessages(
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState {
  switch (action.type) {
    case SEND_INPUT_VALUE: {
      return [...messages, ">>> " + action.payload.value];
    }
    case MESSAGE_JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }
    case MESSAGE_PING: {
      const { server } = action.payload;
      return [...messages, `Received ping from ${server}`];
    }
    case MESSAGE_PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }
    case MESSAGE_NOTICE: {
      return [...messages, action.payload.text];
    }
    default:
      return messages;
  }
}
