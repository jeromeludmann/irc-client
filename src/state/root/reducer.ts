import { RoutedAction, isChannel, isPrivate } from '@app/utils/Route'
import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from '@app/state/route/reducer'
import {
  ServerState,
  reduceServer,
  serverInitialState,
} from '@app/state/server/reducer'
import { CLOSE_WINDOW, ADD_NEW_SERVER } from '@app/actions/ui'

type RootPartialState = Readonly<{
  servers: Readonly<{ [key: string]: ServerState }>
}>

export type RootState = {
  route: RouteState
} & RootPartialState

type RootReducer<S = RootState> = (
  root: S,
  action: RoutedAction,
  extraStates: { root: RootState },
) => S

export const rootInitialState = {
  servers: { serverKey: serverInitialState },
  route: routeInitialState,
}

const caseReducers: { [action: string]: RootReducer<RootPartialState> } = {
  [CLOSE_WINDOW]: (root, action, _) => {
    const thereIsOnlyOneServer = Object.keys(root.servers).length <= 1
    const isChannelOrPrivate =
      isChannel(action.route.bufferKey) || isPrivate(action.route.bufferKey)

    if (thereIsOnlyOneServer || isChannelOrPrivate) {
      return root
    }

    const servers = { ...root.servers }
    delete servers[action.route.serverKey]
    return { servers }
  },
}

const routeActionToServers = (
  servers: { [key: string]: ServerState },
  action: RoutedAction,
  extraStates: { route: RouteState },
) =>
  action.route.serverKey in servers || action.type === ADD_NEW_SERVER
    ? {
        ...servers,
        [action.route.serverKey]: reduceServer(
          servers[action.route.serverKey],
          action,
          extraStates,
        ),
      }
    : servers

export function reduceRoot(
  root: RootState | undefined = rootInitialState,
  action: RoutedAction,
): RootState {
  // prevent other actions to pass inside app reducers
  if (action.route === undefined) {
    return root
  }

  const intermediateState = {
    ...root,
    route: reduceRoute(root.route, action, { root }),
    servers: routeActionToServers(root.servers, action, {
      route: root.route,
    }),
  }

  return {
    ...intermediateState,
    ...(action.type in caseReducers
      ? caseReducers[action.type](intermediateState, action, {
          root: intermediateState,
        })
      : {}),
  }
}
