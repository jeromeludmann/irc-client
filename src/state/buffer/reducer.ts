import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from '@app/state/messages/reducer'
import {
  InputState,
  inputInitialState,
  reduceInput,
} from '@app/state/input/reducer'
import { RoutedAction } from '@app/utils/Route'
import { RouteState } from '@app/state/route/reducer'
import { RECEIVE_JOIN, ReceiveJoinAction } from '@app/actions/messages/incoming'
import { SWITCH_WINDOW } from '@app/actions/ui'
import { ServerState } from '@app/state/server/reducer'
import { AnyAction } from 'redux'

type BufferPartialState = Readonly<{
  activity: boolean
}>

export type BufferState = Readonly<{
  input: InputState
  messages: MessagesState
}> &
  BufferPartialState

type BufferReducer<S = BufferState> = (
  buffer: S,
  action: RoutedAction,
  extraStates: {
    route: RouteState
    server: ServerState
  },
) => S

export const bufferInitialState = {
  activity: false,
  input: inputInitialState,
  messages: messagesInitialState,
}

const itIsMe = (action: AnyAction, extraStates: { server: ServerState }) =>
  action.payload.user.nick === extraStates.server.user.nick

const caseReducers: { [action: string]: BufferReducer<BufferPartialState> } = {
  [RECEIVE_JOIN]: (_, action: ReceiveJoinAction, extraStates) => ({
    activity: !itIsMe(action, extraStates),
  }),

  [SWITCH_WINDOW]: () => ({ activity: false }),
}

export const reduceBuffer: BufferReducer = (
  buffer,
  action,
  extraStates,
): BufferState => ({
  ...(action.type in caseReducers
    ? caseReducers[action.type](buffer, action, extraStates)
    : buffer),
  input: reduceInput(buffer.input, action),
  messages: reduceMessages(buffer.messages, action, extraStates),
})
