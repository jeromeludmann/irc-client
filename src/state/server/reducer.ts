import { AnyAction, Action } from 'redux'
import { RouteState } from '@app/state/route/reducer'
import {
  RoutedAction,
  BufferKey,
  isChannel,
  isPrivate,
  isRaw,
} from '@app/utils/Route'
import {
  RECEIVE_RPL_MYINFO,
  ReceiveReplyMyInfoAction,
  RECEIVE_NICK,
  ReceiveNickAction,
  RECEIVE_PART,
  ReceivePartAction,
  RECEIVE_PRIVMSG,
  RECEIVE_JOIN,
} from '@app/actions/messages/incoming'
import {
  BufferState,
  bufferInitialState,
  reduceBuffer,
} from '@app/state/buffer/reducer'
import {
  CLOSE_WINDOW,
  CloseWindowAction,
  UPDATE_SERVER_LAG,
  UpdateServerLagAction,
} from '@app/actions/ui'

export type ServerState = Readonly<{
  name: string
  user: Readonly<{
    nick: string
    user: string
    real: string
  }>
  lag: number
  modes: Readonly<{
    user: string[]
    available: Readonly<{
      channel: string[]
      user: string[]
    }>
  }>
  buffers: Readonly<{
    [key: string]: BufferState
  }>
}>

type ServerReducer<S = ServerState> = (
  server: S,
  action: Action,
  extraStates: { route: RouteState },
) => S

type BufferRouterReducer = (
  buffers: { [key: string]: BufferState },
  action: RoutedAction,
  extraStates: { route: RouteState; server: ServerState },
) => { [key: string]: BufferState }

export const serverInitialState = {
  name: '<unknown>',
  user: {
    nick: 'default_nick',
    user: 'default_user',
    real: 'default_name',
  },
  lag: 0,
  modes: {
    user: [],
    available: { channel: [], user: [] },
  },
  buffers: {
    [BufferKey.RAW]: bufferInitialState,
    [BufferKey.STATUS]: bufferInitialState,
  },
}

const removeCurrentBuffer = (
  buffers: { [key: string]: BufferState },
  bufferKey: string,
) => {
  const bufferMap = { ...buffers }
  delete bufferMap[bufferKey]
  return bufferMap
}

const removeAllServerRelatedBuffers = (buffers: {
  [key: string]: BufferState
}) => {
  const bufferMap = { ...buffers }

  Object.keys(bufferMap).forEach(buffer => {
    if (isChannel(buffer) || isPrivate(buffer)) {
      delete bufferMap[buffer]
    }
  })

  return bufferMap
}

const itIsMe = (server: ServerState, action: AnyAction) =>
  action.payload.user.nick === server.user.nick

const caseReducers: { [action: string]: ServerReducer } = {
  [CLOSE_WINDOW]: (server, action: CloseWindowAction) => ({
    ...server,
    buffers:
      isChannel(action.route.bufferKey) || isPrivate(action.route.bufferKey)
        ? removeCurrentBuffer(server.buffers, action.route.bufferKey)
        : removeAllServerRelatedBuffers(server.buffers),
  }),

  [RECEIVE_NICK]: (server, action: ReceiveNickAction) => ({
    ...server,
    user: itIsMe(server, action)
      ? { ...server.user, nick: action.payload.nick }
      : server.user,
  }),

  // We arbitrarily decided to close window when we "/part" the channel.
  // But later, we could make this behavior customizable.
  [RECEIVE_PART]: (server, action: ReceivePartAction) =>
    itIsMe(server, action)
      ? {
          ...server,
          buffers: removeCurrentBuffer(server.buffers, action.payload.channel),
        }
      : server,

  [RECEIVE_RPL_MYINFO]: (server, action: ReceiveReplyMyInfoAction) => ({
    ...server,
    name: action.payload.serverName,
    modes: {
      ...server.modes,
      available: {
        channel: action.payload.availableChannelModes,
        user: action.payload.availableUserModes,
      },
    },
  }),

  [UPDATE_SERVER_LAG]: (server, action: UpdateServerLagAction) => ({
    ...server,
    lag: action.payload.lag,
  }),
}

const broadcastHandlers: { [buffer: string]: BufferRouterReducer } = {
  // stop broadcasting
  [BufferKey.NONE]: buffers => buffers,

  // broadcast to active buffer
  [BufferKey.ACTIVE]: (buffers, action, extraStates) => {
    const key = extraStates.route.bufferKey
    return {
      ...buffers,
      [key]: reduceBuffer(buffers[key], action, extraStates),
    }
  },

  // broadcast to all buffers
  [BufferKey.ALL]: (buffers, action, extraStates) => {
    const toBroadcast: { [key: string]: BufferState } = {}

    Object.keys(buffers).forEach(key => {
      if (!isRaw(key)) {
        toBroadcast[key] = reduceBuffer(buffers[key], action, extraStates)
      }
    })

    return { ...buffers, ...toBroadcast }
  },
}

const routeActionToBuffers: BufferRouterReducer = (
  buffers,
  action,
  extraStates,
) => {
  if (action.route.bufferKey in broadcastHandlers) {
    return broadcastHandlers[action.route.bufferKey](
      buffers,
      action,
      extraStates,
    )
  }

  if (
    action.route.bufferKey in buffers ||
    [RECEIVE_PRIVMSG, RECEIVE_JOIN].indexOf(action.type) > -1
  ) {
    return {
      ...buffers,
      [action.route.bufferKey]: reduceBuffer(
        buffers[action.route.bufferKey],
        action,
        extraStates,
      ),
    }
  }

  return buffers
}

export const reduceServer: ServerReducer = (
  server = serverInitialState,
  action,
  extraStates,
) => {
  const intermediateState = {
    ...server,
    buffers: routeActionToBuffers(server.buffers, action as RoutedAction, {
      ...extraStates,
      server,
    }),
  }

  return {
    ...intermediateState,
    ...(action.type in caseReducers
      ? caseReducers[action.type](intermediateState, action, extraStates)
      : {}),
  }
}
