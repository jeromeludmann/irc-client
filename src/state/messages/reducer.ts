import { Action } from 'redux'
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
  CONNECT_TO_SERVER,
  ConnectToServerAction,
} from '@app/actions/socket'
import {
  ReceiveErrorAction,
  ReceiveJoinAction,
  ReceivePartAction,
  ReceivePrivmsgAction,
  ReceiveNoticeFromServerAction,
  ReceiveNoticeFromChannelAction,
  ReceiveNoticeFromUserAction,
  RECEIVE_ERROR,
  RECEIVE_JOIN,
  RECEIVE_NOTICE_FROM_SERVER,
  RECEIVE_NOTICE_FROM_CHANNEL,
  RECEIVE_NOTICE_FROM_USER,
  RECEIVE_PART,
  RECEIVE_PRIVMSG,
  RECEIVE_PING_FROM_SERVER,
} from '@app/actions/messages/incoming'
import {
  SEND_PONG_TO_SERVER,
  SEND_PRIVMSG,
  SendPrivmsgAction,
} from '@app/actions/messages/outgoing'
import {
  PRINT_HELP_BY_DEFAULT,
  PrintHelpByDefaultAction,
  PRINT_HELP_ABOUT_COMMAND,
  PrintHelpAboutCommandAction,
} from '@app/actions/commands'
import { RouteState } from '@app/state/route/reducer'
import { ServerState } from '@app/state/server/reducer'
import { UPDATE_SERVER_LAG, UpdateServerLagAction } from '@app/actions/ui'

// TODO replace string[] by MessageState[]
export type MessagesState = string[]

type MessagesReducer = (
  messages: MessagesState,
  action: Action,
  extraStates: { route: RouteState; server: ServerState },
) => MessagesState

export const messagesInitialState: MessagesState = []

const caseReducers: { [action: string]: MessagesReducer } = {
  [CONNECT_TO_SERVER]: (messages, action: ConnectToServerAction) => [
    ...messages,
    `Connecting to ${action.payload.host}â€¦`,
  ],

  [CONNECTION_CLOSED]: messages => [
    ...messages,
    'Disconnected from remote host.',
  ],

  [CONNECTION_FAILED]: (
    messages,
    { payload: { message } }: ConnectionFailedAction,
  ) => [...messages, message],

  [PRINT_HELP_BY_DEFAULT]: (
    messages,
    { payload: { commands } }: PrintHelpByDefaultAction,
  ) => [
    ...messages,
    ...Object.keys(commands).map(
      commandName => `${commandName} - ${commands[commandName].description}`,
    ),
    'Type /HELP <command> for more details.',
  ],

  [PRINT_HELP_ABOUT_COMMAND]: (
    messages,
    { payload: { command } }: PrintHelpAboutCommandAction,
  ) => [
    ...messages,
    `Usage: /${command.name} ${command.syntax} - ${command.description}`,
  ],

  [RAW_MESSAGES_RECEIVED]: (messages, action: RawMessagesAction) => [
    ...messages,
    ...action.payload.messages,
  ],

  [RECEIVE_ERROR]: (messages, { payload: { message } }: ReceiveErrorAction) => [
    ...messages,
    message,
  ],

  [RECEIVE_JOIN]: (
    messages,
    { payload: { user, channel } }: ReceiveJoinAction,
  ) => [...messages, `${user.nick} has joined ${channel}`],

  [RECEIVE_NOTICE_FROM_SERVER]: (
    messages,
    action: ReceiveNoticeFromServerAction,
  ) => [...messages, action.payload.text],

  [RECEIVE_NOTICE_FROM_CHANNEL]: (
    messages,
    {
      payload: { user, text },
      route: { bufferKey },
    }: ReceiveNoticeFromChannelAction,
  ) => [...messages, `-${user.nick}/${bufferKey}- ${text}`],

  [RECEIVE_NOTICE_FROM_USER]: (
    messages,
    { payload: { user, text } }: ReceiveNoticeFromUserAction,
  ) => [...messages, `-${user.nick}- ${text}`],

  [RECEIVE_PART]: (
    messages,
    { payload: { user, channel, message } }: ReceivePartAction,
  ) => {
    const baseMsg = `${user.nick} has left ${channel}`
    return [...messages, message ? `${baseMsg} (${message})` : baseMsg]
  },

  [RECEIVE_PING_FROM_SERVER]: messages => [...messages, 'Ping?'],

  [RECEIVE_PRIVMSG]: (
    messages,
    { payload: { user, text } }: ReceivePrivmsgAction,
  ) => [...messages, `${user.nick}: ${text}`],

  [SEND_PONG_TO_SERVER]: messages => [...messages, 'Pong!'],

  [SEND_PRIVMSG]: (messages, action: SendPrivmsgAction, { server }) => [
    ...messages,
    `${server.user.nick}: ${action.payload.text}`,
  ],

  [UPDATE_SERVER_LAG]: (
    messages,
    { payload: { lag } }: UpdateServerLagAction,
    {},
  ) => [...messages, `Server lag: ${lag} ms`],
}

export const reduceMessages: MessagesReducer = (
  messagesState = messagesInitialState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](messagesState, action, extraStates)
    : messagesState
