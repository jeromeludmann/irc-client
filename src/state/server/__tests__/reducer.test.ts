import {
  reduceServer,
  serverInitialState,
  ServerState,
} from '@app/state/server/reducer'
import { closeWindow } from '@app/actions/ui'
import { messageReceivers } from '@app/actions/messages/incoming'
import { User } from '@app/utils/Message'
import { BufferKey } from '@app/utils/Route'
import { bufferInitialState } from '@app/state/buffer/reducer'

describe('reduce server state', () => {
  const initialState: ServerState = {
    ...serverInitialState,
    user: { nick: 'me', user: 'user', real: 'Realname' },
    buffers: {
      ...serverInitialState.buffers,
      '#channel': bufferInitialState,
    },
  }

  const extraStates = {
    route: { serverKey: 'serverKey', bufferKey: '#channel' },
  }

  it('should handle CLOSE_WINDOW on status', () => {
    expect(
      reduceServer(
        initialState,
        closeWindow({
          serverKey: 'serverKey',
          bufferKey: BufferKey.STATUS,
        }),
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should handle CLOSE_WINDOW on channel', () => {
    expect(
      reduceServer(
        initialState,
        closeWindow({ serverKey: 'serverKey', bufferKey: '#channel' }),
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should handle CLOSE_WINDOW on private', () => {
    expect(
      reduceServer(
        initialState,
        closeWindow({ serverKey: 'serverKey', bufferKey: 'nick' }),
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  describe('when it is someone', () => {
    const someone: User = { nick: 'someone', user: 'user', host: 'host' }

    it('should handle RECEIVE_NICK', () => {
      expect(
        reduceServer(
          initialState,
          messageReceivers.NICK('serverKey', someone, ['new_nick']),
          extraStates,
        ),
      ).toMatchSnapshot()
    })

    it('should handle RECEIVE_PART', () => {
      expect(
        reduceServer(
          initialState,
          messageReceivers.PART('serverKey', someone, ['#channel', 'Goodbye!']),
          extraStates,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('when it is me', () => {
    const me: User = { nick: 'me', user: 'user', host: 'host' }

    it('should handle RECEIVE_NICK', () => {
      expect(
        reduceServer(
          initialState,
          messageReceivers.NICK('serverKey', me, ['new_nick']),
          extraStates,
        ),
      ).toMatchSnapshot()
    })

    it('should handle RECEIVE_PART', () => {
      expect(
        reduceServer(
          initialState,
          messageReceivers.PART('serverKey', me, ['#channel', 'Goodbye!']),
          extraStates,
        ),
      ).toMatchSnapshot()
    })
  })

  it('should handle RECEIVE_RPL_MYINFO', () => {
    expect(
      reduceServer(
        initialState,
        messageReceivers['004']('serverKey', 'server', [
          '?',
          '?',
          '?',
          'available_user_modes',
          'available_channel_modes',
        ]),
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should handle UPDATE_SERVER_LAG', () => {
    expect(
      reduceServer(
        initialState,
        messageReceivers.PONG('serverKey', 'server', ['server', 'key']),
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should not broadcast', () => {
    expect(
      reduceServer(
        initialState,
        {
          type: 'UNHANDLED_ACTION',
          route: { serverKey: 'serverKey', bufferKey: BufferKey.NONE },
        },
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should broadcast to active buffer', () => {
    expect(
      reduceServer(
        initialState,
        {
          type: 'UNHANDLED_ACTION',
          route: { serverKey: 'serverKey', bufferKey: BufferKey.ACTIVE },
        },
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should broadcast to all buffers', () => {
    expect(
      reduceServer(
        initialState,
        {
          type: 'UNHANDLED_ACTION',
          route: { serverKey: 'serverKey', bufferKey: BufferKey.ALL },
        },
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should not handle anything', () => {
    expect(
      reduceServer(
        undefined!,
        {
          type: 'WHATEVER',
          route: { serverKey: 'serverKey', bufferKey: '#channel' },
        },
        extraStates,
      ),
    ).toMatchSnapshot()
  })
})
