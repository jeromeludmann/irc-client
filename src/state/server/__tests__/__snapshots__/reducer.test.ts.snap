// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce server state should broadcast to active buffer 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should broadcast to all buffers 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should handle CLOSE_WINDOW on channel 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should handle CLOSE_WINDOW on private 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should handle CLOSE_WINDOW on status 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should handle RECEIVE_PONG_FROM_SERVER 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [
        "Response from server key in -1 ms",
      ],
    },
  },
  "lag": -1,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should handle RECEIVE_RPL_MYINFO 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [
        "a",
        "v",
        "a",
        "i",
        "l",
        "a",
        "b",
        "l",
        "e",
        "_",
        "c",
        "h",
        "a",
        "n",
        "n",
        "e",
        "l",
        "_",
        "m",
        "o",
        "d",
        "e",
        "s",
      ],
      "user": Array [
        "a",
        "v",
        "a",
        "i",
        "l",
        "a",
        "b",
        "l",
        "e",
        "_",
        "u",
        "s",
        "e",
        "r",
        "_",
        "m",
        "o",
        "d",
        "e",
        "s",
      ],
    },
    "user": Array [],
  },
  "name": "?",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should not broadcast 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state should not handle anything 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "default_nick",
    "real": "default_name",
    "user": "default_user",
  },
}
`;

exports[`reduce server state when it is me should handle RECEIVE_NICK 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "new_nick",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state when it is me should handle RECEIVE_PART 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state when it is someone should handle RECEIVE_NICK 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;

exports[`reduce server state when it is someone should handle RECEIVE_PART 1`] = `
Object {
  "buffers": Object {
    "#channel": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [
        "someone has left #channel (Goodbye!)",
      ],
    },
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "me",
    "real": "Realname",
    "user": "user",
  },
}
`;
